
import glob,os
include: "Common.snake"

#BINS=[path.split('/')[1] for path in  glob.glob("binning/Bin_*/Bin_*fasta")]

print(",".join(BINS))

print(",".join(SAMPLES))

print(",".join(GROUPS))

rule all : 
    input: expand("Samples_files/{group}/{bin}/{sample}.cnt.gz",bins=BINS,groups=GROUPS,samples=SAMPLES)

rule bam_readcount:
    input:
        fasta = IN + "/binning/{bin}/{bin}.fasta",
        fai = "Index/{bin}/{bin}.fasta.fai",
        bam_file = "Map/{bin}/{group}/{sample}.sorted.bam",
        bed_file = "Index/{bin}/{bin}.bed"
    output: "Samples_files/{group}/{bin}/{sample}.cnt.gz"
    log:
        "Samples_files/{group}/{bin}/bam_readcount_{sample}.log"
    shell:"""
        samtools index {input.bam_file} {input.bam_file}.bai
        bam-readcount -l {input.bed_file} -f {input.fasta} {input.bam_file} 2> {log} | gzip > {output}
        """

rule Get_Bins_contigs_bed :
  input:  IN + "/binning/{bin}/{bin}.fasta"
  output: bin_bed="Index/{bin}/{bin}.bed"
  shell: "{SCRIPTS}/WriteBed.py {input} > {output}"

rule index_fasta_files:
  input:
    fasta = IN + "/binning/{bin}/{bin}.fasta"
  output:
    fai = "Index/{bin}/{bin}.fasta.fai"
  shell: "samtools faidx {input.fasta}"

rule bowtie_index:
    input:   IN + "/binning/{ref}/{ref}.fasta"
    output:  touch("{path}/{ref}/index.done")
    params:  "{path}/{ref}/index"
    threads: THREADS
    log:     "{path}/{ref}/index.log"
    message: "Building bowtie index for {input}"
    shell:   "bowtie2-build {input} {params} --threads {THREADS} &> {log}"

#consider ignoring reads aligned without their pairs
rule bowtie_align:
    input:   left=left_reads, right=right_reads,
             index="Index/{bin}/index.done"
    output:  "Map/{bin}/{group}/{sample}.ubam"
             #temp("{path}/{group,(sample|group)\d+}.bam")
    params:  flag=lambda w: "-q" if is_fastq(w) else "-f",
             left=IN + "/" + "{sample}/{sample}_R1.fq.gz",
             right=IN + "/" + "{sample}/{sample}_R2.fq.gz",
             index="Index/{bin}/index",
             align="",#"--no-unal --maxins 1000 --n-ceil 0,0 --np 100500",
             view=""#"-q 10"
    threads: THREADS
    log:     "Map/{group}.{sample}.{bin}.bowtie.log"
    message: "Aligning reads of {wildcards.group} onto {params.index} with bowtie"
    #shell:   "bowtie2 -x {params.index} {params.flag} -p {threads} {params.align}"
    #         " -1 {input.left} -2 {input.right} 2> {log} | samtools view -bh {params.view} - > {output}"
    shell:
        'bowtie2 -x {params.index} {params.flag} -p {threads} {params.align} -1 {params.left} -2 {params.right} 2>> {log} | samtools view -bh {params.view} - > {output}'


rule samtools_sort:
    input:
        "{path}/{name}.ubam"
    output:
        "{path}/{name}.sorted.bam"
        #temp("{path}/{name}.sorted.bam")
    threads: THREADS
    log:
        "{path}/{name}.sort.log"
    message: "Sorting {input}"
    run:
        shell("samtools sort --threads %d -T {wildcards.path}/{wildcards.name}.tmp -O bam -o {output} {input} &> {log}" % (threads - 1))

